### PR TITLE
Switch to using the .NET SDK properties that are defined to provide the TFM instead of using the assembly that defines System.Object

### LINKED ISSUE #80621: Interop Source Generator TFM check is very expensive
### Description

A customer VS slowdown has been traced back to the TFM check we do in the interop source generators. In particular, the `compilation.GetSpecialType(SpecialType.System_Object)` call is showing up as very expensive to do on every keystroke. There's [a bug open in Roslyn](https://github.com/dotnet/roslyn/issues/66410) for why code that call utilizes is so slow (from the same trace) as it shouldn't be this slow and there's some suspect algorithms, but we should consider changing how we do the TFM check to something cheaper, especially since the TFM check is only used in this repo as none of the interop source generators support downlevel usage and ship with the TFM.

### Configuration

Windows x64

### Regression?

The underlying issue in Roslyn looks to be a regression, but this is not a regression in the source-generator.

### Data
MicrosoftTeams-image

### Analysis

The call to `compilation.GetSpecialType(SpecialType.System_Object)` in the `CreateStubEnvironment` method.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Switch to using MSBuild properties defined by the .NET SDK or inferring the TFM from the generator version (as it ships with the TFM and is versioned identically) and using another SyntaxProvider to enable us to construct a `StubEnvironment` object accurately without having to query any information directly from the `Compilation` object. In addition to fixing #80621, this change effectively enables all of the calculation work here to be done less often in a VS session. Previously we recalculated all of the information per keystroke. Now we'll only recalculate the `SkipLocalsInit` check when the user provides or removes a `SkipLocalsInit` attribute in their code or changes their TFM.

We should backport this PR to .NET 7 to ensure good performance in VS.

Fixes #80621

### CODE REVIEW

File: src/libraries/Common/tests/SourceGenerators/GlobalOptionsOnlyProvider.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.Interop;
+using Microsoft.Interop.UnitTests;
+using SourceGenerators.Tests;
+
+namespace SourceGenerators.Tests
```
Comment (jaredpar) [inline src/libraries/Common/tests/SourceGenerators/GlobalOptionsOnlyProvider.cs:14]: File scoped namespaces are amazing, you all should think about using them :smile:

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs [hunk 2]
```diff
@@ -5,16 +5,81 @@
...
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
 
 namespace Microsoft.Interop
 {
     public static class IncrementalGeneratorInitializationContextExtensions
     {
+        // This type is a record to get the generated equality and hashing operators
+        // which will be faster than the reflection-based ones.
+        private record struct TargetFrameworkSettings
```
Comment (jaredpar) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:19]:
```suggestion
        private readonly record struct TargetFrameworkSettings
```

File: src/libraries/System.Runtime.InteropServices/tests/Common/TargetFrameworkOptions.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+using Microsoft.Interop.UnitTests;
+using SourceGenerators.Tests;
+
+namespace Microsoft.Interop.UnitTests
+{
+    /// <summary>
+    /// An implementation of <see cref="AnalyzerConfigOptions"/> that provides configuration in code
+    /// of the target framework options. Used when testing interop source generators.
+    /// </summary>
+    public class TargetFrameworkConfigOptions : AnalyzerConfigOptions
```
Comment (jaredpar) [inline src/libraries/System.Runtime.InteropServices/tests/Common/TargetFrameworkOptions.cs:20]:
```suggestion
    public sealed class TargetFrameworkConfigOptions : AnalyzerConfigOptions
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs [hunk 4]
```diff
@@ -5,16 +5,81 @@
...
+            public Version? Version { get; init; }
+        }
+
+        private static readonly Version FirstNonCoreVersion = new(5, 0);
+
+        // Parse from the informational version as that is the only version that always matches the TFM version
+        // even in debug builds.
+        private static readonly Version ThisAssemblyVersion = Version.Parse(
+            typeof(IncrementalGeneratorInitializationContextExtensions).Assembly
+            .GetCustomAttribute<AssemblyInformationalVersionAttribute>().InformationalVersion.Split('-')[0]);
```
Comment (jaredpar) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:41]: Doing a throwing parse in a static initializer always makes me a bit nervous. But if you all have test cases around this i'll feel a bit better.

**Review by jaredpar — APPROVED** [review]

**Comment by jaredpar** [comment]
@chsienki PTAL

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs [hunk 5]
```diff
@@ -5,16 +5,81 @@
...
+
         public static IncrementalValueProvider<StubEnvironment> CreateStubEnvironmentProvider(this IncrementalGeneratorInitializationContext context)
         {
-            return context.CompilationProvider.Select(static (comp, ct) => comp.CreateStubEnvironment());
+            // PERF: With the .NET SDK, this path is equivalent to calling CompilationExtensions.CreateStubEnvironment.
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:45]: The perf argument doesn't seem relevant here as it is likely a bug. Rather the real concern is about Roslyn layering correctness.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs [hunk 6]
```diff
@@ -5,16 +5,81 @@
...
+                    _ => TargetFramework.Framework
+                }, Version: tfm.Version ?? ThisAssemblyVersion));
+
+            var isModuleSkipLocalsInit = context.SyntaxProvider.ForAttributeWithMetadataName(
+                TypeNames.System_Runtime_CompilerServices_SkipLocalsInitAttribute,
+                (node, ct) => node is ICompilationUnitSyntax,
+                // If SkipLocalsInit is applied at the top level, it is either applied to the module
+                // or is invalid syntax. As a result, we just need to know if there's any top-level
+                // SkipLocalsInit attributes. So the result we return here is meaningless.
+                (context, ct) => true)
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:74]:
```suggestion
                    TypeNames.System_Runtime_CompilerServices_SkipLocalsInitAttribute,
                    (node, ct) => node is ICompilationUnitSyntax,
                    // If SkipLocalsInit is applied at the top level, it is either applied to the module
                    // or is invalid syntax. As a result, we just need to know if there's any top-level
                    // SkipLocalsInit attributes. So the result we return here is meaningless.
                    (context, ct) => true)
```

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs [hunk 7]
```diff
@@ -5,16 +5,81 @@
...
 using System.Text;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
 
 namespace Microsoft.Interop
 {
     public static class IncrementalGeneratorInitializationContextExtensions
     {
+        // This type is a record to get the generated equality and hashing operators
+        // which will be faster than the reflection-based ones.
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:18]: >         // which will be faster than the reflection-based ones.

What does this mean? Can't we just define the `Equals`? I don't see how one can be generated any faster. Also, what is the "reflection-based" ones?

Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:18]: I'm referring to how the `System.ValueType`'s base override of `Equals()` does a reflection-based equality check for value types. When using a `record struct`, we get a generated `Equals()` method that is exactly what we would have written manually and is kept up to date by the compiler if we ever change how data is stored in this structure.

File: src/libraries/Common/tests/SourceGenerators/GlobalOptionsOnlyProvider.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/Common/tests/SourceGenerators/GlobalOptionsOnlyProvider.cs:14]: I agree, but I don't feel like fighting that fight today 😆.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs (hunk 4 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:41]: The majority of our test suite is going down the route that will use this member, so it's heavily covered in our testing.

File: src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs (hunk 7 above)
Comment (AaronRobinsonMSFT) [reply src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/IncrementalGeneratorInitializationContextExtensions.cs:18]: I would drop the comment. I'm not convinced it adds clarity and only applies if we don't write it ourselves. The generation of the equality overloads is the reason for the `record` concept so I think the benefits are obvious.

**Review by chsienki — APPROVED** [review]
LGTM. An alternative approach would be to use the `MetadataReferencesProvider` to look at `System.Runtime` or appropriate to figure out the target version rather than relying on the TFM from MSBuild; that would also only change when the user actually changes the TFM, rather than per key stroke, but I think either approach is fine.

**Comment by AaronRobinsonMSFT** [comment]
> An alternative approach would be to use the MetadataReferencesProvider to look at System.Runtime or appropriate to figure out the target version rather than relying on the TFM from MSBuild;

@chsienki Would this mean looking at the version of the `System.Runtime` assembly?

**Comment by jkoritzinsky** [comment]
I looked at using the MetadataReferencesProvider. The up-side is that it will work in more scenarios and we won't need to fall back to our "use current assembly version" case as often. However, it looks like it can be a file path or an assembly name, which makes it hard to parse and adds complexity. I think for now, this is the best solution.

**Comment by jkoritzinsky** [comment]
The timeout is a known issue according to Build Analysis, so I'm going to merge this in.
